Support configurable pixel formats for the embedder API sofware rendering backend

ardera opened the pull request:
This PR adds the following things to the embedder API:
- a new kind of backing store `FlutterSoftwareBackingStore2`, which is basically the same as `FlutterSoftwareBackingStore` but with a `struct_size` and `pixel_format` field
- a `FlutterSoftwarePixelFormat` enum, containing a bunch of skia supported pixel formats, like `kGray8`, `kRGBA8888`, `kRGB565`
- some tests that'll just draw solid red/blue/green into a 1x1 backing store and check that the output matches what is expected

fixes https://github.com/flutter/flutter/issues/81332

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.
- [x] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

chinmaygarde commented:
Sorry about the late response. I was on the fence on this one. We recommend that embedder API users use the GPU backends wherever possible. The software backend has many other obvious optimizations that we have not bothered with because we only ever use it in test contexts. So, using the software backend on an already underpowered device makes me nervous. Can you elaborate your use case a bit? Is GPU backend not an option at all? Which boards are you trying to support for example? Maybe we can tinker with them ourselves.

OTOH supporting the most useful formats seems very reasonable. Can we just support popular formats like RGB 565 instead of exposing everything Skia supports?

ardera commented:
> We recommend that embedder API users use the GPU backends wherever possible. The software backend has many other obvious optimizations that we have not bothered with because we only ever use it in test contexts.

Yeah, I mean we could also look into using OpenGL via some kind of software emulation (llvmpipe, swift shader) but using the software rendering skia supports internally seemed like the most straightforward (and also least-overhead) way to go

> Can you elaborate your use case a bit? Is GPU backend not an option at all? Which boards are you trying to support for example?

Sure, concretely I'm working with an [i.MX 6 ULL](https://www.nxp.com/products/processors-and-microcontrollers/arm-processors/i-mx-applications-processors/i-mx-6-processors/i-mx-6ull-single-core-processor-with-arm-cortex-a7-core:i.MX6ULL) for example. That variant of the iMX 6 has no GPU, but sadly it's still used in lots of embedded projects. It does have some hardware acceleration for common 2D graphics operations with a hardware block called the PXP, but that's exposed via a vendor-specific API and skia can't make use of it (without modification). We did some testing and it's able to do 20- maybe 25FPS, but that's with manual conversion of the graphics buffers to RGB565 using NEON SIMD stuff, so maybe 30 might be possible with direct RGB565 rendering?

30FPS (even 20 actually) is good enough, you can't really expect more with that kind of hardware. AFAICT Qt won't do more than that as well.

> OTOH supporting the most useful formats seems very reasonable. Can we just support popular formats like RGB 565 instead of exposing everything Skia supports?

sure, I'll update it

ardera commented:
@chinmaygarde thx for the ping! forgot about it. Removed the 10bit pixel formats

zanderso commented:
@chinmaygarde ping.

ardera commented:
any news on this? @chinmaygarde

zanderso commented:
@chinmaygarde @iskakaushik After a rebase, it looks like this should get a review.

zanderso commented:
From PR triage: The presubmit checks are failing due to a need to rebase.

ardera commented:
> From PR triage: The presubmit checks are failing due to a need to rebase.

ah, accidentally rebased on my forks master instead of upstream master.

ardera commented:
> From PR triage: The presubmit checks are failing due to a need to rebase.

thx, rebased it

iskakaushik reviewed (changes requested):
left some initial comments, looks good overall

ardera commented on shell/platform/embedder/tests/embedder_test_backingstore_producer.cc:
```diff
@@ -5,25 +5,37 @@
+                      "backing store producer should produce deprecated v1 "
+                      "software backing "
+                      "stores.";
+    std::abort();
```
okay sure, but that code is only executed inside the test cases & it's just a sanity check (basically an assert). There's no way an embedder can reach this code

If you want to move the check to the caller I can add an assert in here: https://github.com/flutter/engine/blob/379199b89c12823f5864ce2b208825deb2a5dad2/shell/platform/embedder/tests/embedder_config_builder.cc#L361-L370

that's the only place this constructor is called with a custom pixel format, but tbh I feel like that check is better off in the ctor. let me know if I should change it anyway

But I noticed there was one other place with an `abort` that was in fact reachable by an embedder (would get triggered if the embedder passes an invalid pixel format) and I fixed that :+1:

ardera commented on shell/platform/embedder/common.h:
```diff
@@ -0,0 +1,42 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Okay, did that

zanderso commented:
From PR review triage: Is this PR ready for another review pass?

ardera commented:
@zanderso I addressed the review comments except for one, but I explained it. Maybe @iskakaushik can tell me if it's okay that way or if I should remove the last `std::abort`s (which can only happen inside test cases) anyway. Other than that it's ready for review again

chinmaygarde commented:
@ardera Any updates on the licenses? Perhaps someone else can carry the patch to completion?

skia-gold commented:
Gold has detected about 12 new digest(s) on patchset 16.
View them at https://flutter-engine-gold.skia.org/cl/github/26995

ardera commented:
@chinmaygarde licenses should be fixed now. though don't know what that gold bot comment is about
